validators/hard_fork.ak

```diff
@@ -0,0 +1,528 @@
+        (quantity == expected_quantity)?,
+      }
+    //
+    // Spend(6) requirement: Output has only master token and ada
+    expect
+      fortuna.value_has_only_master_and_lovelace(out_value, own_validator_hash)?
+    // Check output datum contains correct epoch time, block number, hash, and leading zeros
+    // Check for every divisible by 2016 block: 
+    // - Epoch time resets
+    // - leading zeros is adjusted based on percent of hardcoded target time for 2016 blocks vs epoch time
+    expect InlineDatum(output_datum) = out_datum
+    // Spend(7) requirement: Expect Output Datum to be of type State
+    expect State {
+      block_number: out_block_number,
+      current_hash: out_current_hash,
+      leading_zeros: out_leading_zeros,
+      difficulty_number: out_difficulty_number,
+      epoch_time: out_epoch_time,
+      current_posix_time: out_current_posix_time,
+      interlink: out_interlink,
+      extra,
+    }: State = output_datum
+
+    // Spend(8) requirement: Check output has correct difficulty number, leading zeros, and epoch time
+    expect
+      if block_number % epoch_number == 0 && block_number > 0 {
+        // use total epoch time with target epoch time to get difficulty adjustment ratio
+        // ratio maxes out at 4/1 and mins to 1/4
+        let total_epoch_time =
+          epoch_time + averaged_current_time - current_posix_time
+        let (adjustment_numerator, adjustment_denominator) =
+          fortuna.get_difficulty_adjustment(total_epoch_time)
+        // Now use ratio to find new leading zeros difficulty
+        let (new_difficulty, new_leading_zeroes) =
+          fortuna.get_new_difficulty(
+            difficulty_number,
+            leading_zeros,
+            adjustment_numerator,
+            adjustment_denominator,
+          )
+        //
+        and {
+          (new_leading_zeroes == out_leading_zeros)?,
+          (new_difficulty == out_difficulty_number)?,
+          (0 == out_epoch_time)?,
+        }
+      } else {
+        let new_epoch_time =
+          epoch_time + averaged_current_time - current_posix_time
+        // 
+        and {
+          (leading_zeros == out_leading_zeros)?,
+          (difficulty_number == out_difficulty_number)?,
+          (new_epoch_time == out_epoch_time)?,
+        }
+      }
+    //
+    // Spend(9) requirement: Output posix time is the averaged current time
+    expect (out_current_posix_time == averaged_current_time)?
+    //
+    // Spend(10) requirement: Output block number is the input block number + 1 
+    // Spend(11) requirement: Output current hash is the target hash
+    expect
+      (block_number + 1 == out_block_number && out_current_hash == found_bytearray)?
+    //Spend(12) requirement: Check output extra field is within a certain size
+    expect (builtin.length_of_bytearray(builtin.serialise_data(extra)) <= 512)?
+    // Spend(13) requirement: Check output interlink is correct
+    (fortuna.calculate_interlink(
```

Comment:
Maybe move this into an `and` with the condition from the previous expect and then just remove expect all together from the end?